### ISSUE TITLE
[Feature Request] Make achievements work offline

### ISSUE THREAD

**Issue by crazygamerdude17** [libretro/RetroArch#14718]
Ok so, with the rising popularity of Valve's Steam Deck more and more people are using RetroArch as a way of emulating their favorite retro games on this new handheld, the issue is that many of these people are achievements hunters who also commonly play away from home, the issue with that is the fact that if you're offline then you can't get achievements that you want, many achievements can be missed and if you are offline when the game gives you the achievements you may need to start all over again from ground zero of the game. The way I was thinking this could work is by having RetroArch create an encrypted file (encrypted to prevent users from tempering and adding achievements that they truly did not earn properly) and then read that encrypted file once it detects that the user is back on wifi, when it reads the file it should tell RetroAchievements to give the logged-in user these achievements, have it read the user's account to verify that RetroArch has registered all achievements and then after it has confirmed this user had all achievements in the file, delete the file so there is no bloat from multiple files. Finally, remember, this isn't just with the Steam Deck, more and more people are buying portable computers, many are buying specifically for emulation so this means that offline achievements could be one of the best updates to RetroArch, after all, it could use more features like this, I am on a mission to get as many emulators to allow offline achievements and RetroArch is the first ingredient as it includes all of the main emulation cores, thank you and please, please get this to work

**Comment by sydbarrett74** [comment]
Significant, but not insuperable. I understand if the core team doesn't
want to prioritise it, but the OP should be commended for proposing it, at
least.

**Comment by crazygamerdude17** [comment]
> I think the need for encryption (or otherwise securing the cached/offline achievements from being tampered with) is a pretty significant problem.

I would say let that just be a small amount of protection, the RetroAchievements users are all achievement hunters so most would have no reason to want to cheat, and as for a few cheaters that exist, they wouldn't benefit in any way, disable hardcore achievements in offline mode maybe, or maybe even find a way of checking to see if a change was made by RetroArch, if the change isn't made by RetroArch then don't count it

**Comment by theflyingape** [comment]
IMO, the driver here is that a lot of retrogaming cab/console/handheld don't have Internet. And, there's wonderful RA client framework in place for achievements, why not craft a local web service to act as a hub?

Kids want this to "light up" their experiences from family & friends retrogaming parties, without having it post to a mule account online. And no one should suggest to deter this request that a locally-rendered achievement can be used to attest "earning" the badges harvested & moderated by an Internet service -- that should be too obvious.

**Comment by GrayDragonM** [comment]
I'm posting here because this topic is sort of similar to my request, and I don't think it's good to clutter the issues page, but I'd like to see an offline achievement system that doesn't affect the online achievement system and shows a list in RetroArch.

**Comment by manyhats-mike** [comment]
Stumbled on this issue while searching for a way to have offline achievements.  I have a Steam Deck, going on a cruise.  It's the perfect time to dip into the classics and hunt some achievements, but no internet connection obviously.  I would be 100% ok with having achievements available for me offline/local only.

**Comment by scrufulufugus** [comment]
+1 to this. I had some achievements fail to register since the Retro Achievements site went down while I was playing.

**Comment by anegoda1995** [comment]
+1 to this, because I can't get achievements on my handheld without wifi (that only at home)

**Comment by marimbaboy** [comment]
Maybe the amount of time to make this happen is worthwhile for some kind of a rollout.

**Comment by AltPlus30** [comment]
With basically every handheld device including Steam Deck I can't see why this is not a feature. Although, it's probably not really on Libretro team's side. Even though, it's possible, I think RetroAchievements staff should work on some kind of "offline mode". PSN and Steam already have a system for syncing achievements manually or when internet connection starts working.

**Comment by Jobima1st** [comment]
offline achievements will be nice, but I don't think they will be added as Hardcore

**Comment by ToyKeeper** [comment]
Would be really cool to have offline "softcore" achievements, especially with how popular retro handhelds have gotten lately and how they're designed for playing when not at home.  It should be able to sync to the server whenever the device gets online again, even if that's during a totally separate session.  It's the only way to make achievements usable on a lot of devices.

But RetroAchievements are primarily designed for "hardcore" achievements, where the execution environment is heavily restricted to prevent cheating... and "softcore" is more of an afterthought.  So it seems offline is way harder to implement than it really needs to be, because of restrictions which aren't needed or even relevant in softcore mode.  I hope there will be a solution for this, for people who are seeking casual fun on the go instead of hardcore challenges.

It's complicated though.  RetroArch and RetroAchievements are different projects run by different people, so it'd take some coordination and cooperation in order to make it work.

**Comment by shaneo88** [comment]
> Would be really cool to have offline "softcore" achievements
>

I think this is a good way to do this. Keep hardcore as is and online only and make softcore available offline.

**Comment by gitdine** [comment]
Apparently [this is being worked on](https://www.thememorycore.com/feb-18-2026-offline-retroachievements/?ref=the-memory-core-newsletter), but it will have to be implemented individually by each emulator/core if I understand correctly. Fingers crossed!

**Comment by shaneo88** [comment]
> Apparently [this is being worked on](https://www.thememorycore.com/feb-18-2026-offline-retroachievements/?ref=the-memory-core-newsletter), but it will have to be implemented individually by each emulator/core if I understand correctly. Fingers crossed!

Yes! About an hour after I replied to a comment last night I saw that RA had gotten offline achievements working on MelonDS. Once successful they’ll get them working for other open source emulators. 

Exciting. It’s the main reason I don’t play my Retroid Pocket 5 enough. Most places I go do have wifi, and I don’t want to miss achievements

**Comment by SirPrimalform** [comment]
Well this is great news. As someone who doesn't care about bragging rights at all, the heavy "anti-cheating" restrictions put me off using retro achievements at all. But if I can just cache the achievement set for the game at the beginning and just get them locally I'll be happy. I don't even care whether they get uploaded to the site or not.

**Comment by misantronic** [comment]
I am currently working on an approach to solve this via a proxy. Check it out at https://raofflineproxy.com/

**Comment by Cyborg-Taco** [comment]
Yeah I heard about that looks like they have Linux support now. Really excited to have that added.
